### PR TITLE
Initial SQLite Atlas file format

### PR DESCRIPTION
**Pull request by lucaspcram**
### Description:

These are the beginnings of a SQLite atlas file format. Things are still very much WIP, but I would like to get them merged in to encourage collaboration.

### Potential Impact:

No downstream impact.

### Unit Test Approach:

This is an area we still need to flesh out. Testing has been ad-hoc, using the various creation commands found in the PR.

### Test Results:

Results look good so far, but we initially have only implemented the basics.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

**Comment by lucaspcram** [comment]
@hallahan The PR is up.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/packed/PackedAtlas.java [hunk 1]
```diff
@@ -687,6 +693,45 @@ public void save(final WritableResource writableResource)
...
+            {
+                file = (File) resource;
+            }
+            else
+            {
+                file = new File(java.io.File.createTempFile(resource.getName(), ".db"));
+            }
+
+            // TODO see note in the PackedAtlasFieldDelegator class
+            PackedAtlasFieldDelegator delegator = new PackedAtlasFieldDelegator()
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/packed/PackedAtlas.java]: Nice : )

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import org.openstreetmap.atlas.utilities.compression.IntegerDictionary;
+import org.openstreetmap.atlas.utilities.maps.LongToLongMap;
+
+/**
+ * This delegator allows the {@link SQLiteWriter} to access internal {@link PackedAtlas} fields it
+ * needs to efficiently build the database.
+ *
+ * @author lcram
+ */
+public class PackedAtlasFieldDelegator
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java]: Nit. Typically they say "Delegate", not "Delegator".

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+{
+    // TODO this class has a lot of fields. It may be worth refactoring it out and just using
+    // reflection. Alternatively, we could also just make a ton of PackedAtlas fields public. Let's
+    // discuss these options please.
+
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java]: I think the reason the getters are not public in PackedAtlas is that in theory, you could mutate what gets returned. But, we're not... However, making a copy is also a bad idea, because we don't want to abuse our memory usage.

Personally, I'd just make the getters public in packed atlas, but I'm fine with whatever, as long as we don't do a copy.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+ * To save all of the data within an atlas, we need to get each {@link AtlasItem} type and iterate
+ * though them with inserts into the database. This class declares that a given writer needs to get
+ * the atlas to do its job.
+ *
+ * @author hallahan
+ * @author lcram
+ */
+public class SQLiteItemWriter
+{
+    private static final String INSERT_SQL = "INSERT INTO %s VALUES (?, ?, ?, ?)";
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: I like this

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    private static final String INSERT_TAGS_SQL = "INSERT INTO %s_tags VALUES (?, ?, ?)";
+    private static final String INSERT_RTREE_SQL = "INSERT INTO %s_rtree VALUES (?, ?, ?, ?, ?)";
+
+    // database column constants
+    private static final int FIRST_COLUMN = 1;
+    private static final int SECOND_COLUMN = 2;
+    private static final int THIRD_COLUMN = 3;
+    private static final int FOURTH_COLUMN = 4;
+    private static final int FIFTH_COLUMN = 5;
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: I used to do that, but it get's a bit hard to maintain eventually. ++column is easier. But, I also am fine with whatever.

Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: Would it be preferable to just add a checkstyle exception for this class, specifically for the magic constants rule? That way we could just use integer literals. Or do you still think `++column` is easier to maintain? I did find that you have to be careful about resetting it to `0` all over the place.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java (hunk 2 above)
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java]: Fair enough. Honestly considering changing it to `bean` or `container` or something, since it seems to be straying a bit from the delegate pattern (since it's not delegating so much as it is simply packaging a bunch of references).

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java (hunk 3 above)
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/PackedAtlasFieldDelegator.java]: Yeah, no copies for sure. If this were C/C++, we would have `const` here to help us out, and all parties would be satisfied...

I'm not particularly committed to any of the particular solutions, but I do sympathize with @matthieun 's desire to keep `PackedAtlas` as locked down as possible.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java (hunk 5 above)
Comment (hallahan) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: I have no preference.

File: src/main/java/org/openstreetmap/atlas/proto/adapters/ProtoPackedTagStoreAdapter.java [hunk 6]
```diff
@@ -114,6 +114,7 @@ public ProtoSerializable deserialize(final byte[] byteArray)
         Field indexField = null;
         Long index = -1L;
 
+        // Consider using getters, says nick
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/proto/adapters/ProtoPackedTagStoreAdapter.java]: This is just a note, should we do this in another PR?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    public SQLiteDictionaryWriter(final Connection connection,
+            final PackedAtlasFieldDelegator delegator) throws SQLException
+    {
+        this.connection = connection;
+        this.insertStatement = this.connection.prepareStatement(INSERT_SQL);
+        this.delegator = delegator;
+    }
+
+    public void write(final PackedAtlas atlas) throws SQLException
```
Comment (adahn6) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java]: why does this have an unused parameter?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        // Performance is very bad without doing this!
+        this.connection.setAutoCommit(false);
+
+        final Iterable<? extends AtlasItem> items = getItemIterableOfCorrectType(atlas,
+                this.itemType);
+
+        for (final AtlasItem item : items)
+        {
+            // INSERT feature data
+            final long atlasId = item.getIdentifier();
```
Comment (adahn6) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: this may not be possible and/or efficient, but I think you could add a `getItemSQL()` to each item class, then you could just maintain it with the data itself instead of in a separate set of switches and generic logic. e.g., inside of Node.java, 
```Java
public PreparedStatement getItemSql(Connection connection, String shardName) {
    String itemStatement = connection.prepareStatement(
                 String.format(SQLiteItemWriter.INSERT_SQL, this.itemType.toString().toLowerCase()));
    itemStatement.setLong(SQLiteItemWriter.FIRST_COLUMN, this.getIdentifier());
    itemStatement.setString(SQLiteItemWriter.SECOND_COLUMN, "md5sum_here");
    itemStatement.setString(SQLiteItemWriter.THIRD_COLUMN, shardName);
    itemStatement.setBytes(SQLiteItemWriter.FOURTH_COLUMN,
        new WkbLocationConverter().convert(this.getLocation()));
    return itemStatement;
}
```
Same thing for tag statement, which could be in AtlasEntity.java. Then, you could just iterate over each atlas entity directly, skipping over unrequested / unsupported types, like 
```
for (final AtlasItem item : entitiesIterator) {
    if( checkEntityTypeSupported(item)) {
        item.getItemSQL(connection, shardName).execute(); 
    } 
}
```
Obvs this is scratch code and just a thought! May not be worth the refactor or even actually work.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/Atlas.java [hunk 9]
```diff
@@ -675,6 +675,29 @@
...
+     */
+    void saveAsSQLite(WritableResource resource, boolean buildIndices);
+
+    /**
+     * Save as a SQLite database with matcher
+     *
+     * @param resource
+     *            The SQLite file database to write to
+     * @param buildIndices
+     *            * Whether the database should have the indices built for you.
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/Atlas.java]: Nit: extra *

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/AtlasDirectorySQLiteConverter.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+        if (overwrite)
+        {
+            try
+            {
+                FileUtils.deleteDirectory(sqliteDirectory.toFile());
+            }
+            catch (final IOException noDelete)
+            {
+                logger.warn(
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/AtlasDirectorySQLiteConverter.java]: Log the `noDelete`, just in case.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/AtlasDirectorySQLiteConverter.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * This CLI is similar to SQLiteConverter, except here you give it a directory to search for
+ * atlases. You also give it a SQLite directory where those atlas shards will then be converted into
+ * their equivalent SQLite database shard.
+ *
+ * @author hallahan
+ */
+public class AtlasDirectorySQLiteConverter extends Command
+{
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/AtlasDirectorySQLiteConverter.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        for (int key = 0; key < size; ++key)
+        {
+            int column = 0;
+            final String value = dictionary.word(key);
+            this.insertStatement.setInt(++column, key);
+            this.insertStatement.setString(++column, value);
+            this.insertStatement.execute();
+        }
+
+        this.connection.commit();
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java]: Are we leaking connections? This is never closed anywhere. Possibly implement `AutoCloseable` here.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+        if (item.getType() != this.itemType)
+        {
+            throw new CoreException(
+                    "ItemType {} of AtlasItem {} did not match expected ItemType {}",
+                    item.getType(), item, this.itemType);
+        }
+
+        switch (item.getType())
+        {
+            case NODE:
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: Nit: To simplify, consider breaking this down into `LineItem` and `LocationItem` since `Edge`/`Line` processing is the identical, as is `Point`/`Node` processing.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMergeCommand.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            logger.info("Took {} to merge the databases.", mergeTime.elapsedSince());
+
+            if (buildIndices)
+            {
+                logger.info("Building indices in output database...");
+                final Time indicesTime = Time.now();
+                merger.buildIndices();
+                logger.info("Took {} to build indices.", indicesTime.elapsedSince());
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMergeCommand.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMergeCommand.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            }
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMergeCommand.java]: Same

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMerger.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * SQLiteMerger takes a path to an output database file that it will create. It then can attach
+ * Atlas SQLite shard databases and merge those into this output database.
+ *
+ * @author hallahan
+ */
+public class SQLiteMerger
+{
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMerger.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMerger.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    public void buildIndices() throws SQLException
+    {
+        try
+        {
+            this.schema.buildIndices(this.connection);
+        }
+        catch (final SQLException sqlException)
+        {
+            logger.error("Unable to build indices on merge database.");
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMerger.java]: Consider logging the error, might help in debugging

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteSchema.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    private static final String USER_VERSION_TEMPLATE = "PRAGMA user_version = %s";
+    private static final String CACHE_SIZE_TEMPLATE = "PRAGMA cache_size = %s";
+
+    private final String tables;
+    private final String indices;
+
+    /**
+     * The constructor loads the schema file of the matching schema version we are using.
+     */
+    SQLiteSchema() throws IOException, NullPointerException
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteSchema.java]: This is pretty rare to see - throwing NPE. Thoughts on IllegalArgument instead - NPE may lead people down wrong path.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteSchema.java [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+                    .getFile();
+            this.tables = FileUtils.readFileToString(new java.io.File(fileName),
+                    StandardCharsets.UTF_8);
+
+            resourceName = String.format(INDICES_FILE_NAME_TEMPLATE, SCHEMA_VERSION);
+            fileName = Objects.requireNonNull(getClass().getClassLoader().getResource(resourceName))
+                    .getFile();
+            this.indices = FileUtils.readFileToString(new java.io.File(fileName),
+                    StandardCharsets.UTF_8);
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteSchema.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteSchema.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        catch (final IOException noSchema)
+        {
+            logger.error("Unable to read SQLite schema file in resources.", noSchema);
+            throw noSchema;
+        }
+    }
+
+    void buildIndices(final Connection connection) throws SQLException
+    {
+        try (Statement statement = connection.createStatement())
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteSchema.java]: Nice use of try with resources.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteWriter.java [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * @author hallahan
+ */
+public class SQLiteWriter
+{
+    private static final Logger logger = LoggerFactory.getLogger(SQLiteWriter.class);
+
+    private static final String CONNECTION_URL_TEMPLATE = "jdbc:sqlite:%s";
+
+    private final Connection connection;
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteWriter.java]: Connection never closed.

File: src/main/java/org/openstreetmap/atlas/proto/adapters/ProtoPackedTagStoreAdapter.java (hunk 6 above)
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/proto/adapters/ProtoPackedTagStoreAdapter.java]: "says Nick" :) :)

File: src/main/java/org/openstreetmap/atlas/streaming/resource/FileSuffix.java [hunk 22]
```diff
@@ -26,7 +26,8 @@
     TEXT(".txt"),
     ZIP(".zip"),
     WKT(".wkt"),
-    WKB(".wkb");
+    WKB(".wkb"),
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/streaming/resource/FileSuffix.java]: Love it.

**Comment by adahn6** [comment]
@MikeGost changes addressed!

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java (hunk 12 above)
Comment (mgcuthbert) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java]: To add to this comment, I would be very wary of using connections the way you are here by including it as a class variable. I am not sure if there is a need for a connection pool, however you should have a class (like a connection pool) to handle the connections and dole them out to functions as requested. Otherwise you are holding onto that connection the whole time this class is in memory.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+    private final ItemType itemType;
+
+    SQLiteItemWriter(final Connection connection, final boolean buildIndices,
+            final PackedTagStore tagStore, final LongToLongMap identifierToTagStoreIndex,
+            final ItemType itemType) throws SQLException
+    {
+        this.tagStore = tagStore;
+        this.identifierToTagStoreIndex = identifierToTagStoreIndex;
+        this.itemType = itemType;
+        this.connection = connection;
```
Comment (mgcuthbert) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: Again generally not best practice to store connections as class variables.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+                unwrappedStatement.setDouble(THIRD_COLUMN, maxLon);
+                unwrappedStatement.setDouble(FOURTH_COLUMN, minLat);
+                unwrappedStatement.setDouble(FIFTH_COLUMN, maxLat);
+
+                unwrappedStatement.execute();
+            }
+        }
+
+        this.connection.commit();
+        this.connection.setAutoCommit(true);
```
Comment (mgcuthbert) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: All statements would need to be closed after they have been executed/used.

Comment (mgcuthbert) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: Generally best approach is to use them in a try block like 
`try (PreparedStatement unwrappedStatement = this.rtreeStatementOptional.get())
{
  //do something with the statement
}
`
Because the PreparedStatement is AutoCloseable it will clean up everything for you.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+
+            if (keyIndices.length != valueIndices.length)
+            {
+                throw new CoreException(
+                        "key and value array length mismatch, something went wrong");
+            }
+
+            for (int index = 0; index < keyIndices.length; index++)
+            {
+                this.tagsStatement.setLong(FIRST_COLUMN, atlasId);
```
Comment (mgcuthbert) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: This is one of the cases where I would argue that Magic Numbers should be ignored in the check style.

File: src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql [hunk 26]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE metadata (key text NOT NULL, value);
+CREATE TABLE dictionary (key integer PRIMARY KEY, value);
+CREATE TABLE node (id integer PRIMARY KEY, checksum text, shard text, geom blob NOT NULL);
+CREATE TABLE node_tags (id integer NOT NULL, key integer NOT NULL, value integer NOT NULL);
+CREATE TABLE edge (id integer PRIMARY KEY, checksum text, shard text, geom blob NOT NULL);
```
Comment (mgcuthbert) [inline src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql]: I would recommend change the shard column to integer. You can generate shard id's by taking Zoom-X-Y and just removing the dashes and making it an integer. This will give you a performance gain if using it in your where clause especially if you have a index on it.

Comment (mgcuthbert) [reply src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql]: Also for edges I would suggest storing start and end node identifiers.

File: src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql [hunk 27]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE metadata (key text NOT NULL, value);
+CREATE TABLE dictionary (key integer PRIMARY KEY, value);
+CREATE TABLE node (id integer PRIMARY KEY, checksum text, shard text, geom blob NOT NULL);
```
Comment (mgcuthbert) [inline src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql]: You might want to considering storing incoming and outgoing edge identifiers for a node either in an array column or in another table. Although if you store it in a separate table that other table could get very large very quickly.

**Comment by mgcuthbert** [comment]
I added a couple of comments, one thing I would strongly suggest is using a connection pool library like HikariCP to handle your connections for you. I think it would make your life a lot easier.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMerger.java [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+        try (Statement statement = this.connection.createStatement())
+        {
+            // Attach merge database.
+            final String attachSql = String.format(ATTACH_SQL, mergeDatabase.getAbsolutePath());
+            statement.execute(attachSql);
+            logger.info("Attached database {} for merge.", databaseName);
+
+            // Do the merge!
+            statement.executeUpdate(MERGE_EDGES_SQL);
+            statement.executeUpdate(MERGE_EDGES_TAGS_SQL);
```
Comment (mgcuthbert) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteMerger.java]: Is there a reason why we don't merge other types? I don't think I follow exactly how this is merging 2 sqlite database.

**Comment by lucaspcram** [comment]
Thanks for the notes @mgcuthbert . The connections stuff could definitely use some cleanup. However, there have been some scope changes and so we are closing this PR.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java (hunk 23 above)
Comment (hallahan) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteItemWriter.java]: In the past I've done this, but for these purposes, we know we only have one connection to the DB ever. Explicitly closing the connection has no effect on the state of the underlying database file.

File: src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql (hunk 26 above)
Comment (hallahan) [reply src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql]: Yeah, totally. I was thinking about doing both things.

File: src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql (hunk 27 above)
Comment (hallahan) [reply src/main/resources/org/openstreetmap/atlas/sqlite/tables_v1.sql]: For rendering purposes, this can just go in the properties. If this becomes an underlying format for atlas, absolutely! There's a ton of things we can do to optimize the schema and the encoding. For one, the database gets big fast, and some of the compression techniques you see in the current atlas format are good. Like, the Mapquest LineString compression, and the lat lons packed into int64s. SQLite gives you the indexing, but how you store the underlying data being indexed is up to us.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java (hunk 12 above)
Comment (mgcuthbert) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/sqlite/SQLiteDictionaryWriter.java]: For something like SQLite it is a little different because it is in-memory. However it is generally considered best practice to retrieve from a connection pool and then close or return connections as soon as you are done with it. For a shared database the benefits of this kind of practice are very obvious. Connection leaks are much more dangerous and can cause your database to run out of open connections making not only your local application start to fail, but also other applications connecting to the database. A connection pool allows you to limit the connections your application makes and shares an open connection among threads within you application. This gives you a bit of performance so that the application doesn't have to continue opening and closing connections to the database. All these things are not as problematic with something like an in-memory database like SQLite. However if your program grows or even if you start using multiple threads within your application, the same principles apply. 

To answer your "Dumb Question" (and it is in quotes because it really is not a dumb question at all, once the JVM exist all resources associated with the JVM will "close". In this case the socket resource to the database will automatically be shut down. However the connection will not be closed properly. A database can react unexpectedly in that scenario and usually the session in the database will be shutdown in a couple of seconds after the socket closes but there may be cases where the session is kept alive for longer than you would ordinarily want. One such case would be if a SQL query is running at the time the JVM exits, it could cause the session to remain alive for a lot longer than expected, especially if you get into a hairy deadlock kind of situation.

Anyway I understand that in this specific instance connection handling won't have as much impact as it would if you were connecting to a shared database resource, but I would still strongly recommend following best practices so that you won't find yourself in a difficult spot if the code outgrows the initial use cases.
